include/xtensor/xhistogram.hpp

```diff
@@ -57,46 +57,94 @@ namespace xt
-
-        XTENSOR_ASSERT(data.dimension() == 1);
-        XTENSOR_ASSERT(weights.dimension() == 1);
-        XTENSOR_ASSERT(bin_edges.dimension() == 1);
-        XTENSOR_ASSERT(weights.size() == data.size());
-        XTENSOR_ASSERT(bin_edges.size() >= 2);
-        XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
-        XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
-        XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
-
-        xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ bin_edges.size() - 1 });
-
-        auto sorter = xt::argsort(data);
-
-        size_type ibin = 0;
+        return histogramImp(std::forward<E1>(data),
+                            std::forward<E2>(data),
+                            std::forward<E3>(weights),
+                            density,
+                            false);
+    }
 
-        for (auto& idx : sorter)
+    namespace detail
+    {
+        /**
+         * @ingroup histogram
+         * @brief Compute the histogram of a set of data.
+         *
+         * @param data The data.
+         * @param bin_edges The bin-edges. It has to be 1-dimensional and monotonic.
+         * @param weights Weight factors corresponding to each data-point.
+         * @param density If true the resulting integral is normalized to 1. [default: false]
+         * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+         */
+        template <class R = double, class E1, class E2, class E3>
+        inline auto histogramImp(E1&& data, E2&& bin_edges, E3&& weights, bool density = false, bool equal_bins=false)
         {
-            while (data[idx] >= bin_edges[ibin + 1] && ibin < bin_edges.size() - 2)
+            using size_type = common_size_type_t<std::decay_t<E1>, std::decay_t<E2>, std::decay_t<E3>>;
+            using value_type = typename std::decay_t<E3>::value_type;
+
+            XTENSOR_ASSERT(data.dimension() == 1);
+            XTENSOR_ASSERT(weights.dimension() == 1);
+            XTENSOR_ASSERT(bin_edges.dimension() == 1);
+            XTENSOR_ASSERT(weights.size() == data.size());
+            XTENSOR_ASSERT(bin_edges.size() >= 2);
+            XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
+            XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
+            XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
+
+            size_t n_bins = bin_edges.size() - 1;
+            xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ n_bins });
+
+            if (equal_bins)
             {
-                ++ibin;
+                std::array<double, 2> bounds = xt::minmax(data)();
+                double left = bounds[0];
+                double right = bounds[1];
+                double norm = 1. / (right - left);
+                for (size_t i = 0; i < data.size(); ++i)
+                {
+                    auto v = static_cast<double>(data(i));
```

Comment:
Why does `v` needs to be cast to `double`. Couldn't you keep `left`, `right`, and `v` the same value-type as `data`?

Reply:
I agree that casting to double is not a good solution. The problem here is that `value_type` is the type of `weights`, while `data` and `weights` could have different value types. 

Hence, we will need to introduce another `value_type` for `data`, but in pre-calculating of `norm` in the following code,
```
                double norm = 1. / (right - left);
```
We will need take care of the case when `value_type` of `data` is an integer.

Reply:
Maybe you could use `R` (the return type of the funciton, that default to double) as the type of the norm? Otherwise we could also introduce a `get_floating_point_type` metafunctiotaht simply forwards its argument when it's a floating point type, and returns `double` when the argument is an integer.

Reply:
Thanks for the suggestions! I will code it up later to see which one fits better.

Reply:
Now, I am hesitating in front of the histogram code. The return type conversion actually happens at the following line

```cpp
xt::xtensor<R, 1> prob = xt::cast<R>(count);
```

No matter what type `weights` has, it is finally converted to `R`, which is `double` since people usually do not specify the return type `R` in the template. Therefore, we could just make the type of `count` to `R` when declaring it here:

```cpp
xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ n_bins });
```

By the way, I don't understand why only the first overload has the return type `R` in the template.

Reply:
Regarding the type of `norm` as well as `left` and `right`, the point is to calculate the index of the bin `i_bin` accurately. Therefore, in my opinion, converting everything explicitly to `double` may be a good choice. It also suppresses compiler warning.

```cpp
auto i_bin = static_cast<size_t>(static_cast<double>(n_bins) * (static_cast<double>(data(i)) - left) * norm);
```

Reply:
Sorry for the delayed response, I was on holiday. 

I think that having only one overload that allows to specify the return type is simply an oversight. I would say that it should always be consistent. Also disconnecting it from `weights` is the only sensible thing to do when then distribution is normalised to one (`density = true`). 

Then in terms of your question: why is it not allocated to `R`? I think this is simply more efficiently to do the additions with `value_type` in stead of casting for every addition.

I guess I understand the casting strategy for your optimised implementation.

Reply:
> I think this is simply more efficiently to do the additions with value_type in stead of casting for every addition.

Now, I see. And the size of data (number of additions) is usually order of magnitude larger than the number of bins when we need to worry about the performance.Thanks!